### PR TITLE
Add version check caching, fix version comparison

### PR DESCRIPTION
**Pull request by olzhasar-reef**
This PR introduces the following changes:
1. Version checking cache:
Currently, version checking works by making an HTTP request to PyPI on every call. To avoid repetitive network requests, we implement a caching mechanism using a local file, ensuring that requests are only made once a day at most.
2. Improved version comparison:
The current version comparison logic relies on integer comparisons, which do not support letters and may lead to errors in cases like "6.0.15" vs "6.1.0". The new comparison mechanism is more robust and powered by the [version handling](https://packaging.pypa.io/en/latest/version.html) capabilities of the packaging library.
3. Fix for `version_checking` silent failure issue:
The CLI module includes logic for displaying an error message with a tip to disable version checking in case of a network request failure. However, this code is unreachable due to version checking failing silently.
This PR introduces a new `check_version` function with a custom exception. The old behavior is retained with a display of a deprecation warning.
The new function name should also be more descriptive to an end user.

### CODE REVIEW

**Comment by thealligatorking** [comment]
@olzhasar-reef thank you for the contribution!

**Comment by thealligatorking** [comment]
@ibraheem-opentensor do you mind testing this when you get a chance?

File: tests/unit_tests/utils/test_version.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    mock_get_version_from_pypi.assert_called_once()
+    assert version_file_path.exists()
+    assert version_file_path.read_text() == pypi_version
+
+
+@pytest.mark.parametrize("elapsed", [0, VERSION_CHECK_THRESHOLD - 5])
+def test_get_and_save_latest_version_file_fresh_check(
+    mock_get_version_from_pypi: MagicMock, version_file_path: Path, elapsed: int
+):
+    now = datetime.utcnow()
```
Comment (ibraheem-abe) [inline tests/unit_tests/utils/test_version.py]: Thank you for your contribution! 

`utcnow()` is depreciated ([link](https://docs.python.org/3/whatsnew/3.12.html#:~:text=datetime%3A%20datetime.datetime%E2%80%99s%20utcnow()))
Maybe we can use `datetime.now(timezone.utc)`

Comment (olzhasar-reef) [reply tests/unit_tests/utils/test_version.py]: Sure, fixed

**Comment by ibraheem-abe** [comment]
@gus-opentensor  Lgtm!

**Review by thealligatorking — APPROVED** [review]

**Comment by olzhasar-reef** [comment]
@gus-opentensor I've merged recent changes from the staging branch addressing conflicts.
Let me know if there's anything else preventing from merging this PR.

**Pull request merged**